src/server/learnocaml_server.ml

```diff
@@ -176,27 +176,55 @@ module Request_handler = struct
     | Response ({contents; _} as r) -> Response {r with contents = f contents}
     | (Cached _ | Status _) as r -> r
 
+  let alphanum = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+  let alphanum_len = String.length alphanum
+
+  let nonce_req : (Ipaddr.t option, string) Hashtbl.t = Hashtbl.create 533
+
   let token_save_mutex = Lwt_utils.gen_mutex_table ()
 
-  let callback_raw: type resp. string option -> caching -> resp Api.request -> resp ret
-    = fun secret cache -> function
+  let ipaddr_of_endp =
+    function
+    | `TCP (i,_) -> Some i
+    | _ -> None
+
+  let callback_raw: type resp. Conduit.endp -> string option -> caching -> resp Api.request -> resp ret
+    = fun conn secret cache -> function
       | Api.Version () ->
           respond_json cache Api.version
       | Api.Static path ->
           respond_static cache path
+      | Api.Nonce () ->
+         let conn = ipaddr_of_endp conn in
+         begin
+           match Hashtbl.find_opt nonce_req conn with
+           | Some x -> respond_json cache x
+           | None ->
+              let nonce = String.init 20 (fun _ -> alphanum.[Random.int alphanum_len]) in
+              Hashtbl.add nonce_req conn nonce;
+              respond_json cache nonce
+         end
       | Api.Create_token (secret_candidate, None, nick) ->
-         let know_secret =
-           match secret with
-           | None -> true
-           | Some x -> secret_candidate = x in
-         if not know_secret
-         then Lwt.return (Status {code = `Forbidden;
-                                  body = "Bad secret"})
-         else
-           Token.create_student () >>= fun tok ->
-           (match nick with None -> Lwt.return_unit | Some nickname ->
-             Save.set tok Save.{empty with nickname}) >>= fun () ->
-             respond_json cache tok
+         let conn = ipaddr_of_endp conn in
+         begin
+           let forbid s = Status {code = `Forbidden; body = s} in
+           match Hashtbl.find_opt nonce_req conn with
```

Comment:
This code makes me wonder if you do not need an error monad here to improve readability.

Reply:
I agree, but how do you use an error monad in Ocaml ? This refer to a pretty particular thing in Haskell, and I am not sure how to do it in Ocaml.

Reply:
Nothing really special:
https://gist.github.com/yurug/0d0ab0682a11a91a0648b47c036df8ea

By the way, with ocaml 4.08, we will be able to say: 
``
let (let*) x f = bind x f
``
so that `let* x = t in u` will be equivalent to `bind t (fun x -> u)`.

Reply:
Ah thanks for the example, I see (nice new syntax !). Just to be sure, do you want me to change the whole function (`callback_raw`) to be encapsulated in an error monad (maybe something with `either` in place of `option`, since we want to keep the string and the code of the status when failing) ?

Reply:
Yes. Every change that could improve the readability of this code is welcome.

Reply:
Ok, since I changed a bit the `response` type to extract to error case and the `callback_raw` function which gives now a result of type `(resp response, error) result Lwt.t` (with `error = (Cohttp.Code.status_code * string)`.

I defined a useful bind : `( >?= ) : 'a result Lwt.t -> ('a -> 'b result Lwt.t) -> 'b result Lwt.t`, with a strange name to avoid name collision with`Lwt.>>=`.

The whole thing seems more readable to me (even if a do-notation would be welcome :) )